### PR TITLE
Improve SDK to support state param and additional query params passed by webapp

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
Note $subject
Resolves https://github.com/asgardeo/asgardeo-java-oidc-sdk/issues/28

### CODE REVIEW

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/SSOAgentConstants.java [hunk 1]
```diff
@@ -70,6 +70,8 @@
     public static final String HTTP_CONNECT_TIMEOUT = "httpConnectTimeout";
     public static final String HTTP_READ_TIMEOUT = "httpReadTimeout";
     public static final String HTTP_SIZE_LIMIT = "httpSizeLimit";
+    public static final String STATE_PARAM = "state";
+    public static final String ADDITIONAL_PARAMS_FOR_OAUTH2_AUTHZ_ENDPOINT = "additionalParamsForAuthorizeEndpoint";
```
Comment (ChamathNS) [inline io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/SSOAgentConstants.java:74]: Better if we can shorten the constant name. 
suggestion: CUSTOM_PARAMS_FOR_AUTHZ_EP

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java [hunk 2]
```diff
@@ -81,16 +83,32 @@ public OIDCRequestBuilder(OIDCAgentConfig oidcAgentConfig) {
...
+                        .state(state)
+                        .endpointURI(authorizationEndpoint)
+                        .nonce(nonce);
+        // Add additional query params for authentication endpoint.
+        if (oidcAgentConfig.getAdditionalParamsForAuthorizeEndpoint() != null) {
+            for (Map.Entry<String, String> additionalQueryParam : oidcAgentConfig
+                    .getAdditionalParamsForAuthorizeEndpoint().entrySet()) {
+                authenticationRequestBuilder
+                        .customParameter(additionalQueryParam.getKey(), additionalQueryParam.getValue());
+            }
```
Comment (ChamathNS) [inline io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java:108]:
```suggestion
            oidcAgentConfig.getAdditionalParamsForAuthorizeEndpoint().forEach((key, value) -> {
                authenticationRequestBuilder.customParameter(key, value);
            });
```

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java [hunk 3]
```diff
@@ -81,16 +83,32 @@ public OIDCRequestBuilder(OIDCAgentConfig oidcAgentConfig) {
...
         URI callBackURI = oidcAgentConfig.getCallbackUrl();
         URI authorizationEndpoint = oidcAgentConfig.getAuthorizeEndpoint();
-        State state = generateStateParameter();
+        String stateParam = oidcAgentConfig.getState();
+        State state;
+        if (StringUtils.isBlank(stateParam)) {
+            state = generateStateParameter();
+        } else {
+            state = new State(stateParam);
+        }
```
Comment (ChamathNS) [inline io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java:92]: shall we extract this out to a separate private method like `resolveState(oidcAgentConfig)` and call it here?

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java [hunk 4]
```diff
@@ -81,16 +83,32 @@ public OIDCRequestBuilder(OIDCAgentConfig oidcAgentConfig) {
...
-                .nonce(nonce)
-                .build();
+        AuthenticationRequest.Builder authenticationRequestBuilder =
+                new AuthenticationRequest.Builder(responseType, authScope,
+                        clientID, callBackURI)
+                        .state(state)
+                        .endpointURI(authorizationEndpoint)
+                        .nonce(nonce);
+        // Add additional query params for authentication endpoint.
+        if (oidcAgentConfig.getAdditionalParamsForAuthorizeEndpoint() != null) {
```
[the remark this answers is not in the document]
Comment (AnuradhaSK) [reply io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java:103]: If the map is empty, that will be handled by the followed for-loop. So additional check we need to have is the null check. So, adding `MapUtils.isNotEmpty()` check duplicates the map empty check right?

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java (hunk 3 above)
Comment (AnuradhaSK) [reply io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java]: fixed in e7e548c

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java (hunk 2 above)
Comment (AnuradhaSK) [reply io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java]: fixed in e7e548c

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/SSOAgentConstants.java (hunk 1 above)
Comment (AnuradhaSK) [reply io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/SSOAgentConstants.java]: fixed in e7e548c

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java [hunk 5]
```diff
@@ -81,16 +83,32 @@ public OIDCRequestBuilder(OIDCAgentConfig oidcAgentConfig) {
         Scope authScope = oidcAgentConfig.getScope();
         URI callBackURI = oidcAgentConfig.getCallbackUrl();
         URI authorizationEndpoint = oidcAgentConfig.getAuthorizeEndpoint();
-        State state = generateStateParameter();
+        String stateParam = oidcAgentConfig.getState();
```
[the remark this answers is not in the document]
Comment (AnuradhaSK) [reply io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java]: fixed in 6eb0c97

File: io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java [hunk 6]
```diff
@@ -137,6 +151,18 @@ public OIDCRequestBuilder(OIDCAgentConfig oidcAgentConfig) {
...
+    private State resolveState() {
+
+        String stateParam = oidcAgentConfig.getState();
+        State state;
+        if (StringUtils.isBlank(stateParam)) {
+            state = generateStateParameter();
+        } else {
+            state = new State(stateParam);
+        }
+        return state;
```
Comment (ChamathNS) [inline io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java:163]:
```suggestion
        if (StringUtils.isBlank(stateParam)) {
            return generateStateParameter();
        } 
        return new State(stateParam);
```

Comment (AnuradhaSK) [reply io.asgardeo.java.oidc.sdk/src/main/java/io/asgardeo/java/oidc/sdk/request/OIDCRequestBuilder.java]: Fixed in f9781db

**Review by ChamathNS — APPROVED** [review]

**Pull request merged**